File: services/backboard_service.py
"""
Backboard Service
-----------------
Reads and writes the per-user taste profile.

NOTE: The official Backboard SDK (pip install backboard-sdk) exposes
  client.add(user_id, content) / client.add_message(..., memory="Auto")
for memory storage. The REST endpoints below are a best-effort guess;
if they return errors the service falls back gracefully to an in-memory
cache so the pipeline keeps working without remote persistence.
"""

import httpx
from models.schemas import TasteProfile
from config.settings import settings

# In-memory cache to avoid redundant reads within a session
_profile_cache: dict[str, TasteProfile] = {}
_dwell_counts: dict[str, int] = {}


def _headers() -> dict:
    return {
        "Authorization": f"Bearer {settings.BACKBOARD_API_KEY}",
        "Content-Type": "application/json",
    }


def _profile_key(user_id: str) -> str:
    return f"itrack:profile:{user_id}"


# ── Read ──────────────────────────────────────────────────────────────────────

async def get_profile(user_id: str) -> TasteProfile:
    """
    Fetches the user's taste profile from Backboard.
    Falls back to empty profile if not found (first-time user).
    Uses in-memory cache for speed within a session.
    """
    if user_id in _profile_cache:
        return _profile_cache[user_id]

    try:
        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{settings.BACKBOARD_BASE_URL}/memory/{_profile_key(user_id)}",
                headers=_headers(),
                timeout=5.0,
            )
            if resp.status_code == 200:
                data = resp.json()
                profile = TasteProfile(**data.get("value", {}))
                _profile_cache[user_id] = profile
                return profile
    except Exception as e:
        print(f"[Backboard] get_profile failed for {user_id}: {e}")

    # First-time user — return empty profile
    empty = TasteProfile()
    _profile_cache[user_id] = empty
    return empty


# ── Write ─────────────────────────────────────────────────────────────────────

async def update_profile(user_id: str, signals: dict) -> TasteProfile:
    """
    Merges new taste signals (from Gemini extraction) into the existing profile,
    then writes back to Backboard.

    Signals dict shape (from gemini_service.identify_product):
      {
        "style_signals": [...],
        "color_signals": [...],
        "estimated_price_range": "...",
        "brand_guess": "...",
        "product_category": "..."
      }
    """
    profile = await get_profile(user_id)

    # Merge — keep recent 10 unique values per list
    def merge_list(existing: list, new_items: list, limit: int = 10) -> list:
        combined = new_items + existing
        seen = []
        for item in combined:
            if item and item.lower() != "unknown" and item not in seen:
                seen.append(item)
        return seen[:limit]

    profile.preferred_styles = merge_list(profile.preferred_styles, signals.get("style_signals", []))
    profile.preferred_colors = merge_list(profile.preferred_colors, signals.get("color_signals", []))

    brand = signals.get("brand_guess", "")
    if brand and brand.lower() != "unknown" and brand not in profile.preferred_brands:
        profile.preferred_brands = ([brand] + profile.preferred_brands)[:10]

    price = signals.get("estimated_price_range", "")
    if price and price.lower() != "unknown":
        profile.price_range = price

    category = signals.get("product_category", "")
    if category:
        profile.recent_interests = merge_list(profile.recent_interests, [category], limit=5)

    # Write back to Backboard
    try:
        async with httpx.AsyncClient() as client:
            await client.post(
                f"{settings.BACKBOARD_BASE_URL}/memory/{_profile_key(user_id)}",
                headers=_headers(),
                json=profile.model_dump(),
                timeout=5.0,
            )
    except Exception as e:
        print(f"[Backboard] update_profile failed for {user_id}: {e}")
        # Non-fatal — profile is updated in memory even if write fails

    # Update cache
    _profile_cache[user_id] = profile
    _dwell_counts[user_id] = _dwell_counts.get(user_id, 0) + 1

    return profile


def get_dwell_count(user_id: str) -> int:
    return _dwell_counts.get(user_id, 0)

File: src/services/backboardService.ts
import { settings } from "../config/settings.js";
import { TasteProfileSchema } from "../models/schemas.js";
import type { TasteProfile } from "../models/schemas.js";

const profileCache = new Map<string, TasteProfile>();
const dwellCounts = new Map<string, number>();

export interface GeminiSignals {
  product_name: string;
  product_category: string;
  style_signals: string[];
  color_signals: string[];
  estimated_price_range: string;
  brand_guess: string;
}

const mergeList = (existing: string[], incoming: string[], limit: number = 10): string[] => {
  const combined = [...incoming, ...existing];
  const seen = new Set<string>();
  const result: string[] = [];

  for (const item of combined) {
    if (!item) {
      continue;
    }
    if (item.toLowerCase() === "unknown") {
      continue;
    }
    if (!seen.has(item)) {
      seen.add(item);
      result.push(item);
    }
    if (result.length >= limit) {
      break;
    }
  }

  return result;
};

const authHeaders = (): HeadersInit => ({
  Authorization: `Bearer ${settings.BACKBOARD_API_KEY}`,
  "Content-Type": "application/json",
});

export const getProfile = async (userId: string): Promise<TasteProfile> => {
  const cached = profileCache.get(userId);
  if (cached) {
    return cached;
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5000);

  try {
    const response = await fetch(
      `${settings.BACKBOARD_BASE_URL}/memory/itrack:profile:${userId}`,
      {
        method: "GET",
        headers: authHeaders(),
        signal: controller.signal,
      },
    );

    if (response.status === 200) {
      const data: unknown = await response.json();
      const value =
        typeof data === "object" && data !== null && "value" in data
          ? (data as { value: unknown }).value
          : data;
      const parsed = TasteProfileSchema.parse(value);
      profileCache.set(userId, parsed);
      return parsed;
    }
  } catch {
    // Fall back to empty profile below.
  } finally {
    clearTimeout(timeout);
  }

  const empty = TasteProfileSchema.parse({
    preferred_styles: ["minimalist", "streetwear"],
    preferred_colors: ["black", "white"],
    recent_interests: ["sneakers", "outerwear"],
    preferred_brands: [],
    price_range: "$50-$200",
  });
  profileCache.set(userId, empty);
  return empty;
};

export const updateProfile = async (
  userId: string,
  signals: GeminiSignals,
): Promise<TasteProfile> => {
  const current = await getProfile(userId);

  const merged: TasteProfile = {
    ...current,
    preferred_styles: mergeList(current.preferred_styles, signals.style_signals, 10),
    preferred_colors: mergeList(current.preferred_colors, signals.color_signals, 10),
    preferred_brands: current.preferred_brands,
    recent_interests: current.recent_interests,
  };

  const brand = signals.brand_guess;
  if (brand && brand.toLowerCase() !== "unknown" && !merged.preferred_brands.includes(brand)) {
    merged.preferred_brands = [brand, ...merged.preferred_brands].slice(0, 10);
  }

  const priceRange = signals.estimated_price_range;
  if (priceRange && priceRange.toLowerCase() !== "unknown") {
    merged.price_range = priceRange;
  }

  const category = signals.product_category;
  if (category) {
    merged.recent_interests = mergeList(merged.recent_interests, [category], 5);
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5000);
  try {
    await fetch(`${settings.BACKBOARD_BASE_URL}/memory/itrack:profile:${userId}`, {
      method: "POST",
      headers: authHeaders(),
      body: JSON.stringify(merged),
      signal: controller.signal,
    });
  } catch (error) {
    const isNetworkError =
      error instanceof TypeError &&
      typeof (error as { cause?: { code?: unknown } }).cause?.code === "string" &&
      ["ENOTFOUND", "ECONNREFUSED", "ECONNRESET"].includes(
        (error as { cause: { code: string } }).cause.code,
      );
    if (isNetworkError) {
      const code = (error as { cause: { code: string } }).cause.code;
      console.warn(
        `[Backboard] Cannot reach ${settings.BACKBOARD_BASE_URL} (${code}) — profile saved in-memory only. Check BACKBOARD_BASE_URL in your .env.`,
      );
    } else {
      console.warn("[Backboard] updateProfile write failed", error);
    }
  } finally {
    clearTimeout(timeout);
  }

  profileCache.set(userId, merged);
  dwellCounts.set(userId, (dwellCounts.get(userId) ?? 0) + 1);

  return merged;
};

export const getDwellCount = (userId: string): number => {
  return dwellCounts.get(userId) ?? 0;
};

export const clearProfile = (userId: string): void => {
  profileCache.delete(userId);
  dwellCounts.delete(userId);
};
